### ISSUE TITLE
Proposal: Public APIs for C# 8 async streams

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#32640]
We're at a point where we should start exposing the necessary library support for C# 8 async streams, aka async enumerables, aka async iterators.  While we may find ourselves wanting to make further changes to these APIs, we've locked on what we plan to deliver initially, and we can adapt as necessary based on feedback.

```C#
namespace System
{
    public interface IAsyncDisposable
    {
        ValueTask DisposeAsync();
    }
}

namespace System.Collections.Generic
{
    public interface IAsyncEnumerable<out T>
    {
        IAsyncEnumerator<T> GetAsyncEnumerator();
    }

    public interface IAsyncEnumerator<out T> : IAsyncDisposable
    {
        ValueTask<bool> MoveNextAsync();
        T Current { get; }
    }
}

namespace System.Runtime.CompilerServices
{
    public struct AsyncIteratorMethodBuilder
    {
        public static AsyncIteratorMethodBuilder Create();

        public void MoveNext<TStateMachine>(ref TStateMachine stateMachine)
            where TStateMachine : IAsyncStateMachine;

        public void AwaitOnCompleted<TAwaiter, TStateMachine>(
            ref TAwaiter awaiter, ref TStateMachine stateMachine)
            where TAwaiter : INotifyCompletion
            where TStateMachine : IAsyncStateMachine;

        public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(
            ref TAwaiter awaiter, ref TStateMachine stateMachine)
            where TAwaiter : ICriticalNotifyCompletion
            where TStateMachine : IAsyncStateMachine;

        public void Complete();
    }
}
```

**Assembly:**
We need to decide where we put these.  My expectation is .NET Core we'd have all of this in System.Private.CoreLib, but we could also put it in ~~System.Threading.Tasks.Extensions~~ a new assembly.  If we want to OOB this, it could be in both, with the .NET Core ~~S.T.T.Extensions~~ the new assembly just type forwarding to S.P.CoreLib as happens today for ValueTask.

**A few notes:**
- **Alternative `WaitForNextAsync`+`TryGetNext` current design**.  We spent a lot of time exploring this attractive alternative.  It has perf advantages, but also non-trivial additional complexity when using the APIs directly rather than through compiler support (e.g. `foreach await`).  We also have a design where we could light-up with the alternative API should we find that the perf benefits are desirable enough to introduce a second interface.  As such, we're going with the simpler, more familiar, and easier to work with `MoveNextAsync`+`Current` design.
- **Thread safety**. While these APIs are for asynchrony, that doesn't imply an instance of an enumerator can be used concurrently.  Async enumerators are explicitly not thread safe, in terms of there being no way to correctly used `MoveNextAsync`+`Current` in an atomic fashion (without higher-level locking that provided the thread safety), but also in terms of accessing `MoveNextAsync` again before having consumed a previous calls `ValueTask`; doing so is erroneous and has undefined behavior.  So, too, is accessing `DisposeAsync` after having called `MoveNextAsync` but without having consumed its `ValueTask`.
- **`ValueTask` instead of `Task`**.  The original design called for `MoveNextAsync` to return `Task<bool>` and `DisposeAsync` to return `Task`.  That works well when `MoveNextAsync` and `DisposeAsync` complete synchronously, but when they complete asynchronously, it requires allocation of another object to represent the eventual completion of the async operation.  By instead returning `ValueTask<bool>` and `ValueTask`, an implementation can choose to implement `IValueTaskSource` and reuse the same object repeatedly for one call after another; in this fashion, for example, the compiler-generated type that's returned from an iterator can serve as the enumerable, as the enumerator, and as the promise for every asynchronously-completing `MoveNextAsync` and `DisposeAsync` call made on that enumerator, such that the whole async enumerable mechanism can incur overhead of a single allocation.
- **Cancellation**. Cancellation is provided external to the interfaces.  If you want operations on the enumerable to be cancelable, you need to pass a `CancellationToken` to the thing creating the enumerable, such that the token can be embedded into the enumerable and used in its operation.  You can of course choose to cancel awaits by awaiting something that itself represents both the `MoveNextAsync` `ValueTask<bool>` and a `CancellationToken`, but that would only cancel the await, not the underlying operation being awaited.  As for IAsyncDisposable, while in theory it makes sense that anything async can be canceled, disposal is about cleanup, closing things out, freeing resources, etc., which is generally not something that should be canceled; cleanup is still important for work that's canceled. The same CancellationToken that caused the actual work to be canceled would typically be the same token passed to DisposeAsync, making DisposeAsync worthless because cancellation of the work would cause DisposeAsync to be a nop. If someone wants to avoid being blocked waiting for disposal, they can avoid waiting on the resulting ValueTask, or wait on it only for some period of time.
- **Naming**.  `Async` is used in the various method names (e.g. `DisposeAsync`) even though the type also includes `Async` so that a type might implement both the synchronous and asynchronous counterparts and easily differentiate them.
- **`AsyncIteratorMethodBuilder`**.  The compiler could get away with using the existing `AsyncTaskMethodBuilder` or `AsyncVoidMethodBuilder` types, but these both have negative impact that can be avoided by using a new, specially-designed type.  `AsyncTaskMethodBuilder` allocates a `Task` to represent the async method, but that `Task` goes unused in an iterator.  `AsyncVoidMethodBuilder` interacts with `SynchronizationContext`, because `async void` methods need to do so (e.g. calling `OperationStarted` and `OperationCompleted` on the current `SynchronizationContext` if there is one).  And some of the methods are poorly named, e.g. `Start` makes sense when talking about starting an async method, but not when talking about iterating with an iterator.  As such, we introduce a new type tailored to async iterators.  `Create` just returns a builder that the compiler can use, which is likely just `default(AsyncIteratorMethodBuilder)`, but the method might also do additional optional work, like tracing.  `MoveNext` pushes the state machine forward, effectively just calling `stateMachine.MoveNext()`, but doing so with the appropriate handling of `ExecutionContext`.  `AwaitOnCompleted` and `AwaitUnsafeOnCompleted` are exactly what they are on the existing builders.  And `Complete` just serves to notify the builder that the iterator has finished iterating: technically this isn't necessary, and it may just be a nop, but it gives us a hook to be able to do things like tracing/logging should we choose to do so. (We could decide not to include this.)
- **Implementing sync and async interfaces**.  It's fine for a single type to implement both the sync and async counterparts, e.g. `IEnumerable<T>` and `IAsyncEnumerable<T>`.  When consuming manually, the developer can easily distinguish which is being used based on naming (e.g. `GetEnumerator` vs `GetAsyncEnumerator`), and when consuming via the compiler, the compiler will provide syntax for differentiation (e.g. `foreach` vs `foreach await`).

**What else will we want?**
I've opened several additional issues to cover related support we'll want to consider:
- **Extensions for ConfigureAwait** (https://github.com/dotnet/corefx/issues/32684).  The compiler will support pattern-based `foreach await` in addition to binding to the interface, and we can use that to enable the implicit awaits on `MoveNextAsync` and `DisposeAsync` to be done using `ConfigureAwait(false)` by having an extension method like `public static ConfiguredAsyncEnumerable<T> ConfigureAwait<T>(this IAsyncEnumerable<T> enumerable, bool continueOnCapturedContext)`, where the returned `ConfiguredAsyncEnumerable<T>` will propagate that through to a `ConfiguredAsyncEnumerator<T>`, and its `MoveNextAsync` and `DisposeAsync` will return `ConfiguredValueTaskAwaitable`s.
- **ManualResetValueTaskSource** (https://github.com/dotnet/corefx/issues/32664).  If `MoveNextAsync` and `DisposeAsync` were defined to return `Task`s, then the compiler could use `TaskCompletionSource<T>` to implement those async operations.  But as it's returning `ValueTask`, and as we're doing so to enable object reuse, the compiler will be using its own implementation of `IValueTaskSource`.  To greatly simplify that and to encapsulate all of the relevant logic, we should productize the `ManualResetValueTaskSource`/`ManualResetValueTaskSourceLogic` helper type from https://github.com/dotnet/corefx/blob/master/src/Common/tests/System/Threading/Tasks/Sources/ManualResetValueTaskSource.cs.
- **IAsyncDisposable implementations** (https://github.com/dotnet/corefx/issues/32665).  With `IAsyncDisposable` exposed, we'll want to implement on a variety of types in coreclr/corefx where the type could benefit from having an asynchronous disposable capability in addition to an existing synchronous ability.

cc: @jcouv, @MadsTorgersen, @jaredpar, @terrajobst, @tarekgh, @kouvel

**Comment by tarekgh** [comment]
would be nice if you include some sample code for the usage.

**Comment by stephentoub** [comment]
e.g.
```C#
static async IAsyncEnumerable<TResult> Zip<TFirst, TSecond, TResult>(
    IAsyncEnumerable<TFirst> first, IAsyncEnumerable<TSecond> second, Func<TFirst, TSecond, TResult> resultSelector)
{
    using async (IAsyncEnumerator<TFirst> e1 = first.GetAsyncEnumerator())
    using async (IAsyncEnumerator<TSecond> e2 = second.GetAsyncEnumerator())
    {
        while (await e1.MoveNextAsync() && await e2.MoveNextAsync())
            yield return resultSelector(e1.Current, e2.Current);
    }
}
```

**Comment by jnm2** [comment]
Too late to include these APIs in .NET Standard 2.1?

**Comment by onovotny** [comment]
@stephentoub Did we determine what was happening with Ix.NET Async from dotnet/reactive? Especially with the Impl here:

https://github.com/dotnet/reactive/pull/423

Are you taking that code and putting it here in corefx? What about the existing package, etc? We should likely have a discussion on this?

**Comment by stephentoub** [comment]
> Are you taking that code and putting it here in corefx? What about the existing package, etc? We should likely have a discussion on this?

When we've (you, me, @MadsTorgersen, etc.) talked about this previously, my understanding of the plan was that it would continue to live in reactive, and it would be updated based on the actual APIs shipping in the platform.  I don't believe anything has changed there.

**Comment by jaredpar** [comment]
@jnm2 

> Too late to include these APIs in .NET Standard 2.1?

Yes. Particularly because the associated language feature isn't complete yet, and isn't intended to be complete until the .NET Core 3.0 time frame. Need to ship the API + language feature at the same time so they can evolve together.

**Comment by onovotny** [comment]
@stephentoub That's fine with me, it just wasn't clear if that was still the case since it has been a while. Once the core interfaces/compiler work is somewhere we can pull from a private feed, we can update that branch to match. Also happy to give direct commit access to any member on the team that needs it.

**Comment by jcouv** [comment]
@stephentoub Shared a code sample above using high-level primitives. Here's a link to a lowered version of such an async-iterator method: https://gist.github.com/jcouv/ae7800985e3a8700bb84c6650d25bb69 
This is still work-in-progress, but it shows how the new builder type would be called.

```C#
public class C
{
    public static async System.Collections.Generic.IAsyncEnumerable<int> M()
    {
        await System.Threading.Tasks.Task.CompletedTask;
        yield return 4;
    }
}
```

**Comment by akarnokd** [comment]
I have a couple of questions:

1) Do I have to defend against reentrant `MoveNextAsync` invocations? I.e., when the `ValueTask<bool>` completes and the continuation would call `MoveNextAsync` again?

2) `DisposeAsync()` must be called after `MoveNextAsync`'s `ValueTask<bool>` completes and before the next `MoveNextAsync` would be invoked, right? If I wanted to implement a `Timeout` operator (with time) over this, how can I stop the main IAsyncEnumerator safely while the current `ValueTask<bool>` is not yet complete?

3) Cancellation must be external (via `CancellationToken` for example). How can I inject it to a specific instance returned by `IAsyncEnumerable.GetAsyncEnumerator`? In other words, how can I cancel individual runs of the `IAsyncEnumerable`? 

4) If the cancellation has to be established on the source generator, how can I compose LINQ-style operators over such source and carry the cancellation support along with it?

5) Is it mandatory to call (and await) `DisposeAsync()` when the `MoveNextAsync`'s `ValueTask<bool>` completes with false, indicating no more values?

6) If the `ValueTask<bool>` completes with false, should I defend against another invocation of `MoveNextAsync` and if so, what should I do: a) keep returning a false `ValueTask`, b) throw an invalid operation exception, c) have the ValueTask failing with the exception?

**Comment by jkotas** [comment]
> we could also put it in System.Threading.Tasks.Extensions. If we want to OOB this, it could be in both, with the .NET Core S.T.T.Extensions just type forwarding to S.P.CoreLib as happens today for ValueTask.

`System.Threading.Tasks.Extensions` shipped inbox on at least one platform. IIRC, the rules for OOBing that we have agreed on some time ago say that you cannot add to the OOB assembly once it shipped inbox on at least one platform.

**Comment by stephentoub** [comment]
> System.Threading.Tasks.Extensions shipped inbox on at least one platform. IIRC, the rules for OOBing that we have agreed on some time ago say that you cannot add to the OOB assembly once it shipped inbox on at least one platform.

Hmm.  I had it in my head that you couldn't modify/augment APIs of any types that had shipped in such assemblies, but I hadn't internalized that you couldn't add additional unrelated types to them, either.  What's the cause of that?  @weshaggard, @terrajobst, is that actually the case such that if we wanted to go partial OOB here we'd need to introduce a new assembly?

**Comment by stephentoub** [comment]
> Do I have to defend against reentrant MoveNextAsync invocations? I.e., when the ValueTask<bool> completes and the continuation would call MoveNextAsync again?

You need to expect that code will run synchronously as part of the completion of the `ValueTask`, so whatever code is completing the `ValueTask` needs to be ok with arbitrary code running as part of that, including another invocation of `MoveNextAsync`.  If it's not ok with that, it needs to move the completion to a place where it is ok with that, which could include queueing it.  For example, `TaskCompletionSource<T>` can be constructed to `RunContinuationsAsynchronously`, so if you were handing back a `ValueTask` that wrapped a TCS `Task`, you could set that option and then it wouldn't matter where you called `{Try}Set{Result/Exception/Canceled}`, at the expense of the continuation getting queued.  I expect/hope that `ManualResetValueTaskSource{Logic}` will provide a similar option in some fashion.

> DisposeAsync() must be called after MoveNextAsync's ValueTask<bool> completes and before the next MoveNextAsync would be invoked, right?

Yes.  An implementation of the interfaces is not guaranteed to be thread-safe, and most will not be.  So you should not make a second asynchronous call on the object while a first is still in progress.

> If I wanted to implement a Timeout operator (with time) over this, how can I stop the main IAsyncEnumerator safely while the current ValueTask<bool> is not yet complete?

You can't:
https://blogs.msdn.microsoft.com/pfxteam/2012/10/05/how-do-i-cancel-non-cancelable-async-operations/
And see my comments on Cancellation in my original post. You can cancel or timeout your action of await'ing, but you can't influence the implementation you're awaiting unless you own it as well or unless it provides a means for doing so, such as accepting a CancellationToken into whatever method/ctor created the async method.  If you cancel/timeout your await, then you shouldn't interact with that object again, and just let it get GC'd and any objects appropriately finalized.

> Cancellation must be external (via CancellationToken for example). How can I inject it to a specific instance returned by IAsyncEnumerable.GetAsyncEnumerator? In other words, how can I cancel individual runs of the IAsyncEnumerable?

You can't.  It'd be part of the enumerable, or else you can choose to cancel between awaits on your own.  The alternative is to take a `CancellationToken` into `MoveNextAsync`, which does not compose well, in particular with language support, and which can add a lot of cost.

> If the cancellation has to be established on the source generator, how can I compose LINQ-style operators over such source and carry the cancellation support along with it?

The cancellation would be built into the source.  Any LINQ operations over it would implicitly inherit that as part of their awaiting on that source.  If you wanted to further inject cancellation into individual query operators, then those query operators should accept a token as well, e.g. `.FooAsync(..., token).BarAsync(..., token)`.

There's a whole discussion of cancellation at https://github.com/dotnet/csharplang/blob/master/proposals/async-streams.md#cancellation.

> Is it mandatory to call (and await) DisposeAsync() when the MoveNextAsync's ValueTask<bool> completes with false, indicating no more values?

It's never mandatory to call `Dispose`, and it's never mandatory to call `DisposeAsync`.  But it's a good idea to.  And just as you'd want to call `Dispose` on an `IEnumerator<T>` today after its `MoveNext` returns false, you'd want to call `DisposeAsync` on an `IAsyncEnumerator<T>` after its `MoveNextAsync` hands back false.

> If the ValueTask<bool> completes with false, should I defend against another invocation of MoveNextAsyncand if so, what should I do: a) keep returning a false ValueTask, b) throw an invalid operation exception, c) have the ValueTask failing with the exception?

Yes, it's valid (though uncommon) to do so. With `IEnumerator<T>`, calling `MoveNext` after you've already reached the end of the iteration is documented to return false.  The same will be true for `MoveNextAsync`.

**Comment by jkotas** [comment]
> I hadn't internalized that you couldn't add additional unrelated types to them, either. What's the cause of that?

The OOB would need to target a netstandard. If it does not target a netstandard, it is platform specific and we do not need unified type identity. It can be one type identity for netfx and different one for netcoreapp, etc. You can ship OOB like that, but it has limited value.

If the OOB targets netstandard, it should better work on all platforms that support netstandard, including .NET Core 2.1 where this shipped inbox. The problem is which implementation are you going to pick when somebody references the OOB in .NET Core 2.1. It cannot be the inbox implementation because of it is missing the new methods; and it cannot be the OOB implementation because of it would introduce second copy of the inbox types.

**Comment by stephentoub** [comment]
Thanks, @jkotas.  That makes sense.  We've made a bit of a mess, haven't we ;)

Ok, so if we decide we need to go OOB for platforms other than .NET Core 3, we'll need to introduce a new assembly.

**Comment by onovotny** [comment]
@stephentoub Shot in the dark, and subject to API/design reviews anyway, but @bartdesmet split up IxAsync into `System.Linq.Async` and `System.Interactive.Async`. 

Perhaps that's a good place for those interfaces?
https://github.com/dotnet/reactive/tree/IxAsyncCSharp8/Ix.NET/Source/System.Linq.Async/System/Collections/Generic

**Comment by stephentoub** [comment]
> Perhaps that's a good place for those interfaces?

Can you clarify what you mean?  The interfaces are going to ship in .NET Core, not as part of an external library.  If you mean if we decide to ship something OOB, sure, we could consider using such a name for an assembly.  Though OOB isn't the main plan right now; I only mentioned it as an off-hand thing... the delivery vehicle here is .NET Core 3.

**Comment by onovotny** [comment]
@stephentoub I guess I mean that the assembly could/should be part of .NET Core. Ix Async is in grey area already and EF Core depends on it today.

**Comment by akarnokd** [comment]
Thanks.

I have a question regarding the `ValueTask`. Is it more efficient to have a constant `ValueTask` of true/false than having a constant `Task.FromResult` of true/false in synchronous return cases of `MoveNextAsync`?

**Comment by stephentoub** [comment]
> I have a question regarding the ValueTask. Is it more efficient to have a constant ValueTask of true/false than having a constant Task.FromResult of true/false in synchronous return cases of MoveNextAsync?

`ValueTask<T>` is just a struct, and constructing it from a `T` just initializes that struct's fields:
https://github.com/dotnet/coreclr/blob/27c848e37e9998142b60e776cf5b5d08a3543fe1/src/System.Private.CoreLib/shared/System/Threading/Tasks/ValueTask.cs#L474-L481
so it's very cheap to just do `return new ValueTask<bool>(true)` or `return new ValueTask<bool>(false)`.  I'd expect that to be as cheap (or cheaper) as reading/copying from a static containing a "cached" version, but you'd need to measure to see for sure.

**Comment by slang25** [comment]
> the delivery vehicle here is .NET Core 3.

Does that mean that the interfaces will be shipped in a separate package and then be shipped in .NET Core 3.0, a bit like what `ValueTuple` was introduced? Or am I confused?

**Comment by jkotas** [comment]
> the interfaces will be shipped in a separate package and then be shipped in .NET Core 3.0, a bit like how ValueTuple (and Span<T>) were introduced

We have found that trying to patch the existing runtimes with new core features via NuGet package like we have done with ValueTuple always results in sub-par experience. We have tried that many times. Based on the discussions about this, I believe we are going to prioritize the long-term sustainability of the platform over patching the existing runtimes with new core features in creative ways.

For this feature, it would mean:

Ship async streams in .NET Core 3.0 only first in its natural place, without introducing a special little assembly for it. You want to be on .NET Core 3.0 to get the first class async stream experience.

Once it ships in .NET Core, these is an option to ship a best-effort standalone NuGet package to provide async streams for earlier .NET versions. This package can be shipped from corefx or even shipped by community (like https://www.nuget.org/packages/AsyncBridge/). You may use this NuGet package if you desperately need async streams on .NET Framework or earlier .NET Core versions and you can live with the sub-par experience that it provides.

**Comment by slang25** [comment]
I remember when async was first introduced and there was the [Microsoft.Bcl.Async](https://www.nuget.org/packages/Microsoft.Bcl.Async/) package which did a similar job to AsyncBridge.

Perhaps the types could be shipped with Ix.Async for those lower "unsupported" TFMs? The community will want to fill the gap, so this seems like a sensible home for it.

**Comment by GSPP** [comment]
> Alternative WaitForNextAsync+TryGetNext

I have seen many hard-core optimizations around tasks and await on this issue tracker. It seems that over time performance has become a significant priority. Maybe it is prudent to directly skip to the fastest design possible.

This should be seen as low-level machinery that is only touched by experts. Normal users use library functions and language support. Similarly it is very rare that normal C# users must touch enumerators manually.

I think it's correct that performance was so much prioritized in all the other measures described.

**Comment by buybackoff** [comment]
>  It's fine for a single type to implement both the sync and async counterparts, e.g. IEnumerable<T> and IAsyncEnumerable<T>.

In my [library](https://github.com/Spreads/Spreads/blob/master/src/Spreads.Core/Interfaces.cs#L48) I already use async streams for a while and implement both `IEnumerator` and `IAsyncEnumerator` so that they could reuse `Current`. I had to change the contract for  `IEnumerator.MoveNext` so that if it returns false the enumerator is still valid and Current is defined as the value as of last true move.

To the point of performance discussion, another contract of `MoveNext` is that:

> when MoveNext() returns false it means that there are no more elements *right now*, 
and a consumer must call `MoveNextAsync()` and await a new element, or spin
and repeatedly call `IEnumerator.MoveNext` when a new element is expected very soon. Repeated calls to MoveNext() could eventually return true.

The idea is that `MoveNext` could return data as fast as possible and all possible optimizations should be done for the synchronous case. Even wrapping a result in `ValueTask`, calling any unneeded method or checking any condition could be slower than just trying to get a value from an internal buffer and return it, preferably inlining everything along the way.

> Finally, the perf benefits are there (at most saving one interface call per element), but they generally pale in comparison to any kind of I/O being done. 

If there are no more values available instantly in memory then some IO is required and all micro-optimizations are already quite meaningless. The design proposed here with `MoveNextAsync: void -> ValueTask<bool>` is much more convenient and additionally allows to write most of the code as if it is synchronous without any special way of handling async. Then adding `await` is trivial.

*Composition* of synchronous enumerators is also much simpler (LINQ vs Ix style) and availability of new data could be propagated via a side channel, while complex composed operations could be micro-optimized without touching async machinery. My `MoveNextAsync` implementation just awaits for a signal that some new data is available in root data sources and then just calls any complex tree of sync calculations that use synchronous `MoveNext`. That would be possible as well with the alternative design with `TryGetNext/WaitForNextAsync`, which kind of does exactly what my contract with combined `IEnumerator/IAsyncEnumerator`, but will be much more complex.

I believe async streams should not try to replace their sync counterparts when performance matters because they will always be slower. Max performance in terms of method inlining and similar micro optimizations should be outside the scope of async streams. They should not allocate and that is *the must* and already done. And they should have a quick path probing if a value is already available but mostly to avoid jumping to thread pool and yielding, especially on Windows with 15 msec thread time slice. Pragmatically there is almost no such new data that could arrive faster than a couple of virtual/interface calls.

**Comment by quinmars** [comment]
First of all I really appreciate that you found a way to keep a simple interface with acceptable performance characteristics. Out of curiosity or nitpicking, shouldn't a `Zip` implementation await the next items in parallel?

```csharp
private static async IAsyncEnumerable<TResult> ZipIterator<TFirst, TSecond, TResult>(IAsyncEnumerable<TFirst> first, IAsyncEnumerable<TSecond> second, Func<TFirst, TSecond, TResult> resultSelector)
{
    using await (IAsyncEnumerator<TFirst> e1 = first.GetAsyncEnumerator())
    using await (IAsyncEnumerator<TSecond> e2 = second.GetAsyncEnumerator())
    {
        while (true)
        {
            var t1 = e1.MoveNextAsync();
            var t2 = e2.MoveNextAsync();
            if (!await t1 | !await t2)
                yield break;
            yield return resultSelector(e1.Current, e2.Current);
       }
    }
}
```

**Comment by stephentoub** [comment]
> Out of curiosity or nitpicking, shouldn't a Zip implementation await the next items in parallel?

It could.  It all depends on the semantics desired.  It also potentially complicates how you await.  For example, I'd argue the implementation should ensure that no work is currently running when the method completes, and also you don't want to invoke DisposeAsync while one of the async operations is in progress, so instead of:
```C#
if (!await t1 || !await t2) break;
```
you'd probably instead want to do the waiting either like:
```C#
bool r1;
try
{
    r1 = await t1;
}
catch
{
    await t2;
    throw;
}
if (!r1 || !await t2) break;
```
or
```C#
await Task.WhenAll(t1.AsTask(), t2.AsTask());
```

**Comment by quinmars** [comment]
Ah, very good point. I haven't thought of that the second task should be kindly awaited before the `DisposeAsync`. Thank you very much for the explanation.

**Comment by MgSam** [comment]
I think all `IAsyncEnumerable` and related code should be in the box and not require the user to go fetch a NuGet package. I think you should strongly consider making the IAsyncEnumerable version of LINQ in the box as well. 

I'm sure you guys have actual usage statistics but in my experience only a tiny fraction of C# devs know about and use Rx or Ix. Now that it's being promoted to having first-class language support, it needs first-class library treatment as well.

**Comment by GSPP** [comment]
Regarding `Zip`, maybe the best of both worlds can be had like this: `Zip` would not process items in parallel. But we add an extension method `IAsyncEnumerable<TResult> Buffer(IAsyncEnumerable<TResult> items, int itemCount)` that tries to pull items as fast as it can into it's internal bounded size buffer. Then, you could use `.Buffer(1)` before using `Zip` so that both input streams will effectively execute in parallel. This `Buffer` method could be a useful helper to optimize performance in certain cases.

**Comment by quinmars** [comment]
@GSPP It seems like the [current reference implementation](https://github.com/dotnet/reactive/blob/6991d850467b6ec07227584c8556f5cfdfd90c11/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/Zip.cs#L96) of `Zip` process the items in parallel.

**Comment by GSPP** [comment]
@quinmars This is a potentially dangerous design. It introduces parallelism to code that otherwise was and looks sequential. Not sure this should be the default.

If `Enumerable.Zip` did the same thing with `Task.Run` it would clearly be considered inappropriate due to safety concerns (and others).

**Comment by stephentoub** [comment]
> It does not execute anything in parallel.

We don't know that, actually. It does queue two operations to run concurrently. Those MoveNextAsync calls could be doing anything, including executing code that mutates shared state.

**Comment by quinmars** [comment]
Sure, a shared state manipulating enumerator is challenging and if a second enumerator may only run sequentielly you are in trouble. It is possible to take care about that with `Zip`, but putting such restriction on LINQ makes other operators impossible, like `Buffer` (as demonstrated by @GSPP), `Window`, `Merge` and many more.

**Comment by stephentoub** [comment]
> but putting such restriction on LINQ makes other operators impossible

This is conflating things.

Every LINQ operator today is completely possible with that constraint; otherwise the IEnumerable implementations there today couldn't be there. Additional operators like a windowing operator are also completely t feasible.

When you start taking about operators whose definition requires concurrency, like one that merges two concurrent sequences based on availability, then of course, concurrency requires concurrency. But that doesn't mean it should be forced upon every other operator.

Introducing such concurrency also has potential negative consequences, or at least policy implications. It can increase resource consumption. It raises questions about exception handling. It has potentially negative consequences related to shared state. Etc.

Note I'm not saying it shouldn't be an option. But IMO it shouldn't be the default. It's a similar to why methods like the base Stream.CopyToAsync don't overlap reads and writes.

**Comment by divega** [comment]
> Introducing such concurrency also has potential negative consequences, or at least policy implications.

FWIW, the fact that System.Interactive.Async currently uses WhenAll to implement “standard” LINQ operators has been a source of pain for EF Core and our customers. We have taken steps to try to serialize the execution of underlying queries, but it plagued with problems like 
https://github.com/aspnet/EntityFrameworkCore/issues/8393.

**Comment by onovotny** [comment]
The reference implementation can certainly change behavior for 3.0. It definitely needs more eyes on it.

**Comment by quinmars** [comment]
I see. So this rule is not necessarily part of the `IAsyncEnumerable` contract but should be followed by the standard LINQ operators. I'm fine with that.

@onovotny If you like, I can prepare some unit tests for `System.Linq.Async` to test for (non-)concurrency.

**Comment by divega** [comment]
> The reference implementation can certainly change behavior for 4.0. It definitely needs more eyes on it.

Good to know, @onovotny. I added a note to that issues to follow up. We are likely going to file something on the interactive extensions repo with more of a formal ask.

**Comment by onovotny** [comment]
@divega Sounds good. To set expectations, we don't have a ton of people actively looking at it right now. Bart did the current reference updates, but I'm not sure where his time is these days. We could use some more direct involvement in updates as things progress. Whether the code stays where it is or moves elsewhere is also up for discussion if that's a concern.

**Comment by akarnokd** [comment]
Is it okay for a consumer to fire-and-forget `IAsyncDisposable.DisposeAsync()`? In a multi-source operator should the `DisposeAsync` call wait until all involved `IAsyncEnumerator.DisposeAsync()` complete? 

For example, let's say I'd want to write an operator that takes two `IAsyncEnumerable`s and relays items from the one which happened to be responding through the `MoveNextAsync`'s task faster the first time (In Rx.NET terminology: `Amb`). In this case, there is no reason to keep the other async sequence open. Thus when the loser's `MoveNextAsync` task completes, I call `DisposeAsync()` promplty. If the winner async sequence happens to end as well, the consumer awaits the `DisposeAsync()` of the operator. Should this task complete only when both the winner's and the loser's  `DisposeAsync` complete?

If they should be both awaited, what should happen in case winner, the loser or both fail? Options: ignore, take the first failure or aggregate failures.

Note that many similar operators have to run the async sequences concurrently, thus the typical async-await consumption doesn't work in my experience. One has to go for `Task.ContinueWith` and sometimes multiple `TaskCompletionSource` instances to trigger resumption.

**Comment by stephentoub** [comment]
> Is it okay for a consumer to fire-and-forget IAsyncDisposable.DisposeAsync()?

I'd discourage it, as I do for most asynchronous operations.

> In a multi-source operator should the DisposeAsync call wait until all involved IAsyncEnumerator.DisposeAsync() complete?

In general, yes.  Otherwise a consumer doesn't know whether the resources used by all of the constituent operations are still in use.

> If they should be both awaited, what should happen in case winner, the loser or both fail? Options: ignore, take the first failure or aggregate failures.

Different APIs take different approaches, and it depends how important it is for all error information to be propagated.  This is largely a decision that's up to the library.

**Comment by akarnokd** [comment]
After playing with the API for about two weeks and implementing dozens of operators, the main trouble I see is the inconvenience of external cancellation channel(s) required.

For example, let's see the `Amb` operator with two sources. The operator wants to relay items from the faster source, disposing the other one. Based on the rules, this has to await the other's `MoveNextAsync` which may take a long time. In addition, when the faster one finishes, the whole setup can only be disposed if the other allows it. The workaround would be to provide two `CancellationTokenSource`s for each async source so that `Amb` can fire-off cancellation and hope the relevant `Task`s will "unblock".

```cs
public static IAsyncEnumerable<T> Amb<T>(
    IAsyncEnumerable<T> firstSource, CancellationTokenSource firstCancel,
    IAsyncEnumerable<T> secondSource, CancellationTokenSource secondCancel
)

var cancel1 = new CancellationTokenSource();
var source1 = GetFirstSource(cancel1.Token);

var cancel2 = new CancellationTokenSource();
var source2 = GetFirstSource(cancel2.Token);

foreach async (var item in Amb(source1, cancel1, source2, cancel2)) {
  // ...
}
```

Now imagine we have tens of sources to `Amb`, `Zip`, `Merge` or an unknown number of sources to `FlatMap`, etc.

I understand the main purpose of the sequential requirement between `MoveNextAsync` and `DisposeAsync`: to protect `Current` from concurrency when the consumer accesses Current while another thread cancels its async enumerator. I don't know how to resolve this with the current design.

The second moderate problem I found so far is the loss of efficiency due to the need to serialize `MoveNextAsync` with itself and with `DisposeAsync`. This is required when there are more than one sources (or time is involved) which may need to run in parallel yet they have to affect the output async enumerator that can be operated on concurrently. This requires 2 nested trampolines per source, one to protect `MoveNextAsync` against reentrance for being potentially completing synchronously and one to protect a concurrent `MoveNextAsync` from being disposed via `DisposeAsync`. These pose at least 4 atomic adds and 1 `ContinueWith` per item as well as 1 `ContinueWith` per dispose (whatever atomics they may do). In addition, there is 1 atomic add per source and 1 `await` to make sure the output async enumerator's `DisposeAsync` waits for all sources to dispose completely. The cost of enqueue/dequeue of items/state is not included.

(*Just for comparison, Reactive Streams that allows more responsive cancellation and has backpressure, can be implemented in a way that the worst case requires 2 atomic operations per item (same trampolining logic); 1 atomic operation per N items (the request/prefetch amount) and 2 atomic operations for dispose cleanup (if at all). Ideally, certain operators can do better when the producer and consumer gets into a balanced streaming where the producer mostly has to do 1 atomic operation per item and the consumer has to do 1 atomic operation when the a "streak" ends. The cost of enqueue/dequeue of items/state is not included.*)

**Comment by stephentoub** [comment]
Thank you @akarnokd for sharing your thoughts.

>  the main trouble I see is the inconvenience of external cancellation channel(s) required.

Yes, this is one of the holes in the model, impacting more advanced use cases.  I actually wrote an email recently to some other folks on the C# language design team about this, and I'm reformatting it here into markdown to share more broadly...

---
_[Email excerpt]_

For reference, here are the interfaces we’re talking about, as we’ve agreed to them thus far:
 ```C#
namespace System.Collections.Generic
{
    public interface IAsyncEnumerable<out T>
    {
        IAsyncEnumerator<T> GetAsyncEnumerator();
    }
 
    public interface IAsyncEnumerator<out T> : IAsyncDisposable
    {
        ValueTask<bool> MoveNextAsync();
        T Current { get; }
    }
}
```
Note the distinct lack of a `CancellationToken` anywhere.

We've talked about all of this before, in some cases many times, but I want to summarize/augment it here.

## Options

I see (and we’ve discussed) three ways we could pass a `CancellationToken` into this system:
1. We change `MoveNextAsync()` to `MoveNextAsync(CancellationToken cancellationToken = default)`.
2. The current POR, which is that any `CancellationToken` needed is passed to / embedded into the `IAsyncEnumerable<T>`, e.g. an argument to the async iterator creating the `IAsyncEnumerable<T>`.
3. We change `GetAsyncEnumerator()` to `GetAsyncEnumerator(CancellationToken cancellationToken = default)`.

### Option 1: `MoveNextAsync(CancellationToken)`
I don’t believe this is viable, and we should take it off the table:
1. The enumeration is logically all part of a single operation, and it’s difficult to come up with scenarios where you would want to change the cancellation owner on each turn of the crank.
2. It can add cost to an implementation, as an implementation is now essentially forced to register and unregister for every `MoveNextAsync` rather than potentially doing it once for the whole operation.  We’ve made that much cheaper in .NET Core than it was in .NET Framework, but there’s still overhead.
3. We lack a mechanism for passing that token provided to `MoveNextAsync` into the body of a C# async iterator, and we’ve been loath to add any cancellation-specific mechanisms into the language.
4. Even if we could pass one into the body from `MoveNextAsync`, it would be difficult for an implementer to employ, as the token could change across any `yield` (we’ve talked about models where `yield` would return the updated `CancellationToken`, but that’s cumbersome for the aforementioned reasons, and also doesn’t address the case of everything before the first `yield`).  It would also impact the code a developer could write, as they would need to be absolutely sure they unregistered from the previous token prior to any `yield`.
5. We would need either a compiler feature or an extension method to funnel a `CancellationToken` into an `await foreach`, since it’s the compiler that’s generating calls to `MoveNextAsync`.
 
### Option 2: Current POR (`CancellationToken` not in the interfaces)
Our current plan of record has some obvious upsides:
1. It leaves cancellation outside of the language’s scope; C# doesn’t know anything about `CancellationToken` and it gets to continue living in blissful unawareness.
2. It matches how developers pass `CancellationToken` into `async` methods today, e.g. today you might write:
```C#
public async Task<T[]> GetResultsAsync(CancellationToken cancellationToken) { … }
```
and tomorrow you could write:
```C#
public async IAsyncEnumerable<T> GetResultsAsync(CancellationToken cancellationToken) { … }
```
3. It’s naturally usable in an `await foreach`, e.g.
```C#
await foreach (T item in GetResultsAsync(cancellationToken)) { … }
```
4. It works well when writing an `IAsyncEnumerable<T>` async iterator, as the token is available in the body of the iterator as with any other parameter, e.g.
```C#
public async IAsyncEnumerable<T> GetResultsAsync(CancellationToken cancellationToken)
{
    while (await FillBuffer(cancellationToken))
        while (_pos < _buffer.Length)
            yield return GetNextResult();
}
```
But of course it has some potential downsides:
1. The `CancellationToken` ends up effectively embedded in the IAsyncEnumerable, which means you can’t change the `CancellationToken` per `GetAsyncEnumerator` call.  That’s fine for:
```C#
await foreach (T item in GetResultsAsync(cancellationToken)) { … }
```
but not fine for:
```C#
private readonly IAsyncEnumerable<T> _resultsEnumerable = GetResultsAsync();
public async IAsyncEnumerable<T> GetFilteredResultsAsync(Predicate<T> filter, CancellationToken cancellationToken)
{
    await foreach (T item in _resultsEnumerable) // Uh oh: no way to pass in CancellationToken
        if (filter(item))
            yield return item;
}
```
2. Developers that have internals access to an async enumerable generator may be able to craft a system where they support tearing off derivative enumerables from other enumerables but with an added `CancellationToken` (similar to how, for example, PLINQ has complete knowledge of the whole enumerable graph and then builds up the execution graph only when enumeration starts, based on its knowledge of all participating operators), but without knowing about such internals, you can’t build a combinator that’s capable of canceling a `MoveNextAsync` call, e.g.
```C#
public static IAsyncEnumerable<(TFirst,TSecond)> ConcurrentZipAsync<TFirst,TSecond>(IAsyncEnumerable<TFirst> first, IAsyncEnumerable<TSecond> second)
{
    await using (IAsyncEnumerator<TFirst> firstEnumerator = first.GetAsyncEnumerator())
    await using (IAsyncEnumerator<TSecond> secondEnumerator = second.GetAsyncEnumerator())
    {
        while (true)
        {
            Task<bool> firstTask = firstEnumerator.MoveNextAsync().AsTask();
            Task<bool> secondTask = secondEnumerator.MoveNextAsync().AsTask();
 
            Task<bool> completed = await Task.WhenAny(firstTask, secondTask);
            Task<bool> other = completed == firstTask ? secondTask : firstTask;
 
            if (await completed)
            {
               if (!await other) break;
               yield return (firstEnumerator.Current, secondEnumerator.Current);
           }
            else
            {
               // Uh oh:
               // We want to cancel `other` here, but don’t have a way to do so.
               // We could simply not await it, but then we’d be leaving the
               // operation running.  We also would then want to avoid disposing
               // the second enumerator, as we don’t want to dispose it / re-enter
               // the state machine while the MoveNextAsync operation is still in flight.
              await other;
               break;
            }
        }
    }
}
```
 
### Option 3: `GetAsyncEnumerator(CancellationToken)`
This approach has some big upsides as well:
1. You can supply a different `CancellationToken` each time you enumerate.  That means you can create the `IAsyncEnumerable<T>`, store it somewhere, and then later on any number of times iterate it with whatever `CancellationToken` you have at the time:
```C#
private readonly IAsyncEnumerable<T> _resultsEnumerable = GetResultsAsync();
public async IAsyncEnumerable<T> GetFilteredResultsAsync(Predicate<T> filter, CancellationToken cancellationToken)
{
    await using (IAsyncEnumerator<T> e = _resultsEnumerable.GetAsyncEnumerator(cancellationToken)) // huzzah!
    {
        while (await e.MoveNextAsync())
        {
            T item = e.Current;
            if (filter(item))
                yield return item;
        }
    }
}
```
2. You can compose with cancellation over arbitrary enumerables someone else provided. This makes it possible to implement combinators like the `ConcurrentZipAsync` that wasn’t fully possible above.
```C#
public static IAsyncEnumerable<(TFirst,TSecond)> ConcurrentZipAsync<TFirst,TSecond>(IAsyncEnumerable<TFirst> first, IAsyncEnumerable<TSecond> second)
{
    var cts = new CancellationTokenSource();
    await using (IAsyncEnumerator<TFirst> firstEnumerator = first.GetAsyncEnumerator(cts.Token))
    await using (IAsyncEnumerator<TSecond> secondEnumerator = second.GetAsyncEnumerator(cts.Token))
    {
       …
             cts.Cancel(); // huzzah!
             try { await other; } catch (OperationCanceledException) { }
             break;
        …
    }
}
```
3. You can use iterators to implement an `IAsyncEnumerator<T>`, which means you can use that iterator support for -ators to implement -ables, e.g.
```C#
public static IAsyncEnumerable<(TFirst,TSecond)> ConcurrentZipAsync<TFirst,TSecond>(IAsyncEnumerable<TFirst> first, IAsyncEnumerable<TSecond> second) =>
    new ConcurrentZipAsyncIterator<TFirst, TSecond>(first, second);
 
private sealed class ConcurrentZipAsyncIterator<TFirst, TSecond> : IAsyncEnumerable<(TFirst, TSecond)>
{
    private readonly IAsyncEnumerable<TFirst> _first;
    private readonly IAsyncEnumerable<TSecond> _second;
 
    internal ConcurrentZipAsyncIterator(IAsyncEnumerable<TFirst> first, IAsyncEnumerable<TSecond> second)
    {
        _first = first;
        _second = second;
    }
 
    public static IAsyncEnumerator<(TFirst,TSecond)> GetAsyncEnumerator(CancellationToken cancellationToken)
    {
        var cts = new CancellationTokenSource();
        using (CancellationTokenSource combined = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, cts.Token))
        await using (IAsyncEnumerator<TFirst> firstEnumerator = _first.GetAsyncEnumerator(combined.Token))
        await using (IAsyncEnumerator<TSecond> secondEnumerator = _second.GetAsyncEnumerator(combined.Token))
        {
            …
        }
    }
}
```
But, of course, it also has some non-trivial downsides (some of which, in particular the language-related support, we’ve discussed before at length):
1. In the last example above, you obviously have to write a lot more boilerplate to implement an `IAsyncEnumerable<T>`.
2. In that same example, you lose out on the optimization that lets you reuse the -able instance as the -ator, since the iterator will create its own instance.
3. There’s no good way to `await foreach` the enumerable with a `CancellationToken`:
```C#
await foreach (T item in GetResultsAsync()) { … } // Uh oh: no place to pass in the CancellationToken to GetAsyncEnumerator
```
Instead, you would need to manually expand the `await foreach`:
```C#
await using (IAsyncEnumerator<T> e = GetResultsAsync().GetAsyncEnumerator(cancellationToken))
{
    while (await e.MoveNextAsync())
    {
        T item = e.Current;
        …
    }
}
```
But we could also address that by exposing a `WithCancellation` extension method that would take advantage of the pattern-based `await foreach`-binding to funnel the supplied `CancellationToken` through to the `GetAsyncEnumerator` call, e.g.
```C#
public static AsyncEnumerableWithCancellation<T> WithCancellation<T>(this IAsyncEnumerable<T> source, CancellationToken cancellationToken) =>
    new AsyncEnumerableWithCancellation<T> { _source = source, _token = cancellationToken };
public struct AsyncEnumerableWithCancellation<T> // does not implement interface
{
    private IAsyncEnumerable<T> _source;
    private CancellationToken _token;
    public IAsyncEnumerator<T> GetAsyncEnumerator() => _source.GetAsyncEnumerator(_token);
}
…
await foreach (var item in source.WithCancellation(cancellationToken)) { … }
```
Alternatively (or in addition), we could add `IAsyncEnumerator<T>` support to `await foreach`, e.g.
```C#
await foreach (T item in GetResultsAsync().GetAsyncEnumerator(cancellationToken)) { … }
```
but that brings with it lifetime issues: is the enumerator disposed or not?  In the case of `await foreach`’ing an enumerable, the enumerator is disposed, because the `await foreach` creates the enumerator and owns it.  But with `await foreach`’ing an enumerator, the `await foreach` is being given the enumerator.  If it chooses not to dispose of it, then code like this potentially leaks:
```C#
await foreach (T item in GetResultsAsync().GetAsyncEnumerator(cancellationToken))
{
    if (FoundWhatIWasLookingFor(item))
        break;
} // Uh oh: enumerator is not disposed
```
And if it chooses to dispose of it, then code like this could end up using disposed objects:
```C#
using (IAsyncEnumerator<T> e = GetResultsAsync().GetAsyncEnumerator(cancellationToken))
{
    // Find first token
    await foreach (T item in e)
        if (IsFirst(item))
            break;
 
    // Find second token
    await foreach (T item in e) // Uh oh: enumerator was disposed
        if (IsSecond(item))
            break;
}
```
which could be very unexpected, though also looks really weird for someone familiar with using `foreach` on enumerables.

4. A developer that writes an iterator directly will have an async enumerable that completely ignores the `CancellationToken` supplied to `GetAsyncEnumerator`.  The best the compiler could do (if it wanted to become aware of `CancellationToken`) would be to do an upfront check for `IsCancellationRequested` before calling back into the state machine, but that’s it… it would be completely ignored in the iterator, e.g.
```C#
await foreach (T item in GetResultsAsync().Select(i => Process(i)).GetAsyncEnumerator(cancellationToken)
{
    …
}
…
public static IAsyncEnumerable<U> SelectAsync<T,U>(this IAsyncEnumerable<T> source, Func<T, ValueTask<U>> func)
{
    await foreach (T item in source) // Uh oh: CancellationToken passed to GetAsyncEnumerator completely ignored
        yield return await func(item);
}
```
Arguably worse, a developer might write the iterator to accept a `CancellationToken` as an argument, and now there’s two different places to pass the `CancellationToken` in, one of which will work, one of which won’t:
```C#
await foreach (T item in GetResultsAsync(cancellationToken1).Select(i => Process(i)).GetAsyncEnumerator(cancellationToken2)
{
    // Uh oh: cancellationToken1 used and cancellationToken2 ignored
    …
}
```
To address those issues, we would need to introduce some kind of language feature to make the `CancellationToken` available inside of the iterator, which could be a non-trivial thing to design and makes C# aware of `CancellationToken`, something we’ve explicitly avoided doing in the past, e.g.
```C#
public static IAsyncEnumerable<U> SelectAsync<T,U>(this IAsyncEnumerable<T> source, Func<T, ValueTask<U>> func)
{
    await foreach (T item in source.GetAsyncEnumerator(iterator.CancellationToken)) // `iterator` as a contextual keyword exposing the argument to MoveNextasync
        yield return await func(item);
}
 ```

## What does all this mean?
 
I think the only holistic solution currently in front of us is to have `GetAsyncEnumerator(CancellationToken cancellationToken = default)` *and* have that token exposed in the body of an IAsyncEnumer*able*<T> iterator via some new keyword:
- It’s the correct answer from a model perspective. It allows you to cancel the thing that should actually be cancelable: the individual enumeration.
- It’s the composable answer.  Providing a `CancellationToken` is done at the time that the actual enumerator is created, and since that enumerator is then responsible for creating sub-enumerators, the `CancellationToken` can be naturally flowed into them.
- Even without additional language features, it fits perfectly fine with iterators in the language when using iterators to implement `IAsyncEnumerator<T>` (rather than `IAsyncEnumerable<T>`).
- It addresses all known scenarios from a functionality perspective.
- Adding this functionality in the future if/when we decided to add better language support for it would be challenging.

We could also do it without the language feature now, and instead just have guidance about the right way to implement `IAsyncEnumerable<T>` when it's exposed from a public library and when using iterators to implement it.
 
Other than that, we’re weighing usability vs completeness: Option 2 is arguably more usable for what I believe are the vast majority use cases (creating iterators whose result is enumerated once, not using complicated combinators, etc.), and hangs together from the perspective of language features, but it’s an incomplete solution.

**Comment by onovotny** [comment]
@stephentoub Thank you for posting that email. I agree 100% with your conclusions for all the same reasons, and I hope the plan can shift in that direction before it's too late.

**Comment by onovotny** [comment]
For downside number 2, you could still have an optimization where the -able can be the -itor for the first one. It's what we do in Ix Async today. The next time GetAsyncEnumerator is called, we clone the state then to create the additional instance. At that point, we could supply and store a different cancellation token.

**Comment by stephentoub** [comment]
> For downside number 2, you could still have an optimization where the -able can be the -itor for the first one.

No, either I miswrote or I think you misunderstood.  If you're implementing the async enumerable by implementing an IAsyncEnumerable class that exposes a GetEnumerator implemented with iterator support (as I did in the referenced example), then you've already constructed your own IAsyncEnumerable, separate from the object the compiler will allocate when calling GetAsyncEnumerable.

**Comment by onovotny** [comment]
@stephentoub I think I'm misunderstanding then. If I have my IAsyncEnumerable class shaped like this:

https://github.com/dotnet/reactive/blob/IxAsyncCSharp8/Ix.NET/Source/System.Interactive.Async/AsyncIterator.cs

The GetEnumerator call returns `this` to avoid an extra allocation the first time. That same call could store a cancellation token that the iterator uses?

Otherwise, I'm not understanding?

**Comment by stephentoub** [comment]
> Otherwise, I'm not understanding?

I'm talking about using iterator support, i.e. `yield return`.

If I implement the whole thing myself, then sure, I can do exactly what the compiler does and return `this` from `GetAsyncEnumerator`.

But if I implement my own `IAsyncEnumerable<T>` class:
```C#
internal sealed class MyAsyncEnumerable<T> : IAsyncEnumerable<T>
{
    ...
}
```
that then uses _iterator support_ to implement `GetAsyncEnumerator`, e.g.
```C#
public async IAsyncEnumerator<T> GetAsyncEnumerator()
{
    ...
    yield return await ...;
    ...
}
```
then I can't do that.

That was the point I was trying to make.  If you want to implement your `IAsyncEnumerable<T>` with an iterator, if we lack the ability to access the `CancellationToken` supplied to `GetAsyncEnumerator` from inside of the iterator, then you can still achieve it with some more boilerplate by writing your own async enumerable that has a GetAsyncEnumerator implemented with iterator support... but in doing so, you can no longer apply the `return this` optimization.

**Comment by onovotny** [comment]
Ah, ok, got it, yes, I wasn't thinking of the generated iterator, thanks.

Is that something that future language or compiler support could optimize?

With that trade-off it still seems like a worthy trade-off to have a complete solution. The current POR (option 2) concerns me deeply as it prevents me from passing `IAsyncEnumerable`'s around in a similar manner as I do `IEnumerable`. The reason comes down to scope and lifetime of the cancellation...all the reasons you articulated.

**Comment by stephentoub** [comment]
> it prevents me from passing IAsyncEnumerable's around in a similar manner as I do IEnumerable

How so?  You have the exact same expressivity between them.  There's no way to cancel a MoveNext call on `IEnumerator<T>`, so how is it different?

**Comment by onovotny** [comment]
It's different because cancellation is a thing for async operations where it's not really for non-async. I don't expect to be able to cancel an IEnumerator, I would expect to control cancellation of an IAsyncEnumerator.

**Comment by jnm2** [comment]
The `ConcurrentZipAsync` example in the Option 2 section doesn't guard against an exception in `await completed`. This could allow the `other` task to be finalized with an unobserved exception. This probably isn't important to the point being made about cancellation.

**Comment by stephentoub** [comment]
> The ConcurrentZipAsync example in the Option 2 section doesn't guard against an exception in await completed

Yes.  This is all email-compiled and email-tested.  I hope we don't focus on that :)

**Comment by akarnokd** [comment]
Just considering the `IAsyncEnumerator` API, an option would be to define a `void Cancel()` method, which should then propagate upwards to the source-like asnyc enumerators. It is then up to the chain and sources to turn them into the relevant `CancellationTokenSource.Cancel()` calls when needed, otherwise it could be left as no-op or pass-through. This way, no registration/deregistration dance would be needed on an async enumerator level just to get the cancellation notification.

I have no idea how the foreach/using could work with it, or if it even should. The closest analogous problem I can remember is from Ix.NET where the `Finally` operator has to intercept the `Dispose` call ending the iteration which is inaccessible in a `yield return` style method implementation afaik. The operator had to be custom, hand-crafted `IEnumerator`.

**Comment by akarnokd** [comment]
Indeed it is not very .NET like. All in all, I think option 2 is the least problematic from my perspective and I'll have to figure out an API to give the user a token in exchange for a async enumerable source:

```cs
zipMany(int count, Func<int, CancellationToken, IAsyncEnumerable<T>> sourceSupplier)
```

**Comment by stephentoub** [comment]
Since everything else here has been done and the only known open issue is around whether we want to change something for cancellation, I've opened a separate issue just for that.
https://github.com/dotnet/corefx/issues/33338

**Comment by stephentoub** [comment]
Closed by https://github.com/dotnet/corefx/pull/33104.

**Comment by akarnokd** [comment]
Last final question about the lifecycle of the enumerator. Is the following legal?

```cs
var source = asyncSource.GetAsyncEnumerator();
if (!shouldWeContinue) {
   await source.DisposeAsync();
   return;
}
```

I.e., not calling `MoveNextAsync` at all but just disposing right away?

**Comment by stephentoub** [comment]
Yup, that's legal.  Just as it is for `IEnumerator<T>`, there's no requirement for `IAsyncEnumerator<T>` that you must call `MoveNext{Async}`.

**Comment by tmds** [comment]
@stephentoub I read your [interesting blog post about ValueTask](https://blogs.msdn.microsoft.com/dotnet/2018/11/07/understanding-the-whys-whats-and-whens-of-valuetask/). It left me with the question: why `MoveNextAsync()` returns a `ValueTask<bool>`. Since `Task<bool>` is cached for completed, what advantages does the `ValueTask` bring here?

**Comment by tmds** [comment]
Answering my own question: this is to be able to reduce allocations when the `MoveNextAsync()` doesn't complete synchronously (re-using the IValueTaskSource).

**Comment by stephentoub** [comment]
> this is to be able to reduce allocations when the MoveNextAsync() doesn't complete synchronously

Corrrect.  In the normal case, we just allocate an object that is an `IAsyncEnumerable<T>`, `IAsyncEnumerator<T>`, `IValueTaskSource<bool>`, and `IValueTaskSource`, and that one object is then what's handed out when you invoke the iterator method, when you call its GetAsyncEnumerator method, when you invoke its MoveNextAsync and it completes asynchronously, and when you invoke its DisposeAsync and it completes asynchronously.  Further, the runtime is generally able to use that same object as the continuation that's stored into an awaited Task or ValueTask.  So you end up with essentially one (maybe two) allocations of overhead for the entire async iteration.

**Comment by SeanFarrow** [comment]
I apologise if this is handled in another issue, but I'm unable to find anything!

Is there a list of existing BCL types that will get API's returning IAsyncEnumerables in .Net Core 3? I'm thinking of things such  as File.ReadAllLines etc?

**Comment by stephentoub** [comment]
> Is there a list of existing BCL types that will get API's returning IAsyncEnumerables in .Net Core 3? I'm thinking of things such as File.ReadAllLines etc?

Very few. Right now just a new method in the channels library was added, and I would not expect any others in 3.0.

**Comment by SeanFarrow** [comment]
Ok, are there going to be examples to show how to do things like read from streams using IAsyncEnumerable?
Questions that come to mind are things like should I use a StreamReader as per .Net Core 2.x and yield return, or should I derive from IAsyncEnumerable.

Is there any public discussion around what will be added in future?

**Comment by stephentoub** [comment]
> Questions that come to mind are things like should I use a StreamReader as per .Net Core 2.x and yield return, or should I derive from IAsyncEnumerable.

IAsyncEnumerable isn't any different in this regard from IEnumerable.  If you would find the programming model of using an iterator to access lines from the stream valuable, then you'd expose it via `I{Async}Enumerable`.

> Is there any public discussion around what will be added in future?

We're not planning to proactively add a ton of `IAsyncEnumerable` implementations, and will instead be demand driven here.  If there's a particular API you'd like to see, please feel free to propose something by opening an issue.
https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md

> Ok, are there going to be examples to show how to do things like read from streams using IAsyncEnumerable?

Like this?
```C#
static async IAsyncEnumerable<string> ReadLinesAsync(StreamReader reader)
{
    string line;
    while ((line = await reader.ReadLineAsync().ConfigureAwait(false)) != null)
    {
        yield return line;
    }
}
...
await foreach (string line in ReadLinesAsync(reader))
{
    ...
}
```

**Comment by SeanFarrow** [comment]
Thanks, feel free to close this.

**Comment by quinmars** [comment]
> Like this?
> 
> ```cs
> static async IAsyncEnumerable<string> ReadLinesAsync(StreamReader reader)
> {
> ```

Will `IAsyncEnumerable` returning methods also share the `XxxxAsync` naming convention athough they are not awaitable?

**Comment by stephentoub** [comment]
Yes. The most common consumption will be via `await foreach`, and it serves as a similar visual indicator of where asynchrony is involved, allows for differentiation from sync enumerables, etc.

**Comment by atykhyy** [comment]
Converting my existing APIs to use async streams, quite a lot of places have cropped up where an `IAsyncEnumerable<T>` is a poor fit semantically because it has to represent the result of a one-time operation. I.e. once `GetAsyncEnumerator()` has been called on the new enumerable and used for enumeration, the underlying source of items is consumed and cannot be enumerated any more. Simple example: an asynchronous parser that parses a `Stream` into MIME parts. Nothing says that the stream has to be rewindable, and in fact usually it isn't. The choices seem to be either to have the method return an `IAsyncEnumerable` and document that it is not reusable, or return an `IAsyncEnumerator` which can be `await foreach`ed only by means of wrapping it in a dummy struct implementing `IAsyncEnumerable`. This feels like a nasty kludge to express a simple intent.

This problem has knock-on effects on converters or processors that asynchronously enumerate items of one kind and convert/join/split them into items of a different kind. If they take an `IAsyncEnumerable`, the resulting enumerable may or may not be reusable depending on what is passed to the converter, which may not be obvious to the caller. If they take an `IAsyncEnumerator`, callers cannot use `await foreach` without a kludge.

The issue of how to supply the cancellation token that is discussed in #33338 and the LDM discussion linked from it also appears to be connected to the distinction between an asynchronous enumeration that can be repeated as many times as necessary (in this case supplying the cancellation token to each particular instance of the enumeration via `GetAsyncEnumerator` is logical, whereas supplying one to the method creating the enumerable doesn't make much sense) and a one-shot enumeration, where the situation is reversed. If users are allowed to use `IAsyncEnumerator` to express this intent, the issue will be mitigated and the two cancellation token problem can be avoided completely.

My ideal solution would be to allow `await foreach` on `IAsyncEnumerator`s. The lifetime question brought up above is important, but it could be handled in two ways: (a) minor syntax modification, something like `await foreach (var t in using EnumerateItemsAsync (...))` or `await using foreach (var t in EnumerateItemsAsync (...))` with perhaps a warning message if `await foreach` without `using` is used on a return value rather than a local variable to prevent common error patterns; or (b) a `ConfigureAwait`-like generic extension method that would return a struct wrapping the original enumerator object and only exposing `MoveNextAsync` and `Current`, preventing `await foreach` from seeing the `IAsyncDisposable` and disposing the enumerator.

A potentially more lightweight alternative might be to allow the compiler to pick up extension `GetAsyncEnumerator` methods as is already done for `GetAwaiter`. This would permit use of `await foreach` on `IAsyncEnumerator`s if a dummy extension method similar to

    public static IAsyncEnumerator<T> GetAsyncEnumerator<T> (this IAsyncEnumerator<T> enumerator) => enumerator ;

is in scope. The lifetime question can be handled along the lines of option (b) above.

**Comment by atykhyy** [comment]
Having played for a week with the extension-method approach that I'd called kludgy above, it actually feels quite natural, and is not especially burdensome compared to `IAsyncEnumerable` because in library code every enumeration of an `IAsyncEnumerable` needs a `.ConfigureAwait(false)` tacked on, too. These same extension methods let me express explicitly whether a given `await foreach` has to dispose the enumerator or not. I have made a [NuGet package](https://github.com/elasticbytes/async-enumerable) with these extension methods, and retracted my syntax proposals above.
